h2o-core/src/main/java/hex/AUC2.java

```diff
@@ -263,30 +265,34 @@ public void perRow(double pred, int act, double w ) {
       idx = -idx-1;             // Get index to insert at
 
       // If already full bins, try to instantly merge into an existing bin
-      if( _n > _nBins ) {       // Need to merge to shrink things
+      if (_n == _nBins &&
+              _speed &&                    // Optimization enabled
+              idx > 0 && idx < _n &&       // Give up for the corner cases
+              _ths[idx - 1] != _ths[idx])  // Histogram has duplicates (mergeOneBin will get rid of them)
+      {       // Need to merge to shrink things
         final int ssx = find_smallest();
-        double dssx = compute_delta_error(_ths[ssx+1],k(ssx+1),_ths[ssx],k(ssx));
-
+        double dssx = _sqe[ssx] + _sqe[ssx+1] + compute_delta_error(_ths[ssx+1], k(ssx+1), _ths[ssx], k(ssx));
         // See if this point will fold into either the left or right bin
         // immediately.  This is the desired fast-path.
-        double d0 = compute_delta_error(pred,w,_ths[idx  ],k(idx  ));
-        double d1 = compute_delta_error(_ths[idx+1],k(idx+1),pred,w);
-        if( d0 < dssx || d1 < dssx ) {
-          if( d1 < d0 ) idx++; else d0 = d1; // Pick correct bin
-          double oldk = k(idx);
-          if( act==0 ) _fps[idx]+=w;
-          else         _tps[idx]+=w;
-          _ths[idx] = _ths[idx] + (pred-_ths[idx])/oldk;
-          _sqe[idx] = _sqe[idx] + d0;
-          assert ssx == find_smallest();
+        double d0 = _sqe[idx-1] + compute_delta_error(pred,w,_ths[idx-1],k(idx-1));
+        double d1 = _sqe[idx] + compute_delta_error(_ths[idx],k(idx),pred,w);
+        if (d0 < dssx || d1 < dssx) {
+          if (d0 <= d1) idx--; // Pick correct bin
+          if (ssx == idx-1 || ssx == idx)
```

Comment:
I believe you are correct in this case.  It should be the total sqe generated by merging this point and just the extra error by merging pred into one of the bins.

Reply:
It is also important for the floating point math (for comparison with the old version where everything relied on mergeOneBin).